File: 2646-kth-largest-sum-in-a-binary-tree/kth-largest-sum-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public long kthLargestLevelSum(TreeNode root, int k) {

        // here also we use levelorder traveersal like leetcode 199
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        List<Long> levelSum = new ArrayList<>();

        while(!queue.isEmpty()) {
            int n = queue.size();

            long sum = 0;

            for(int i = 0 ; i<n ; i++) {
                TreeNode node = queue.poll();
                sum = sum + node.val;

                if(node.left!=null) {
                    queue.add(node.left);
                }
                if(node.right!=null) {
                    queue.add(node.right);
                }
            }

            levelSum.add(sum);

        }


        // sort list with decresing order
        Collections.sort(levelSum);
        Collections.reverse(levelSum);

        // If k is valid, return the kth largest level sum
        if (k <= levelSum.size()) {
            return levelSum.get(k - 1);
        }

        // If k is invalid, return -1 or handle as per problem requirement
        return -1;
    }
}

File: 2646-kth-largest-sum-in-a-binary-tree/kth-largest-sum-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long long kthLargestLevelSum(TreeNode* root, int k) {
        if (!root) return -1;

        queue<TreeNode*> q;
        q.push(root);

        vector<long long> levelSum;

        while (!q.empty()) {
            int n = q.size();
            long long sum = 0;

            for (int i = 0; i < n; ++i) {
                TreeNode* node = q.front();
                q.pop();
                sum += node->val;

                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }

            levelSum.push_back(sum);
        }

        // Sort in descending order
        sort(levelSum.begin(), levelSum.end(), greater<long long>());

        if (k <= levelSum.size()) {
            return levelSum[k - 1];
        }

        return -1;
    }
};
